### DISCUSSION TITLE
Allow BitmapFont / BitmapText to have higher resolution than application

### DISCUSSION THREAD

**Discussion by xzilja** [pixijs/pixijs#7212] [Ideas]
I was playing with performance tweaks of my app that runs on @3x devices.  In general all images look perfectly fine rendered at resolution of 2 and the only thing that seems blurry and off is text (I use BitmapFont and BitmapText).

I tried setting my `PIXI.Application` resolution to 2 and `BitmapFont.from` resolution in options to 3, but it seems it is still using application one.

Would it be possible / is it possible to at some time allow rendering text at higher resolutions than rest of the app somehow?

**Comment by bigtimebuddy** [comment]
I thought this was possible? Did you try this (the options after style).

```js
BitmapFont.from('fontname', {}, { resolution: 2 });
```

**Reply by xzilja** [reply]
Yep, I tried that and `BitmapFont` resolution can't be higher than application one, it can only be lower. See examples below. What I am trying to achieve is sharp text in example 1, but it is not working. And I assume resolution option for `BitmapFont` is set and working correctly as per example 2, where it is set to be different than apps one.

Just a note, I used 1 and 2 values in examples below to exaggerate the effect, but as per my question above, they should be 2 and 3 (2 for application, 3 for text)

1. Here is application at 1x resolution and `BitmapFont` at 2x

<img width="300" alt="Screenshot 2021-01-30 at 09 18 33" src="https://user-images.githubusercontent.com/3154053/106350156-69114a00-62dc-11eb-8d53-da6d355fc827.png">

 ---

 2. Application at 2x resolution and `BitmapFont` at 1x
 
<img width="300" alt="Screenshot 2021-01-30 at 09 19 10" src="https://user-images.githubusercontent.com/3154053/106350174-81816480-62dc-11eb-9390-c1023e66d790.png">

---

3. Everything at 2x

<img width="300" alt="Screenshot 2021-01-30 at 09 21 54" src="https://user-images.githubusercontent.com/3154053/106350200-a675d780-62dc-11eb-9f12-bd01b9c801f3.png">

**Reply by xzilja** [reply]
I also edited the title, to reflect what I am proposing better.

**Comment by bigtimebuddy** [comment]
Okay, I see what's going on. What you're looking to do is to set a low resolution on Renderer/Application, but selectively have sharp assets (BitmapText). I'm not sure your approach here is correct.

The Application resolution defines the amount of pixels/size of the canvas. Even if you set elements at a higher resolution, the pixels that are rendered to stay the same (lower). In fact, increasing the resolution on something like Text or BitmapFonts only increases your memory footprint without actually looking any better.

You should set your Application renderer at the _maximum_ resolution and then down-res items selectively for better performance.

**Reply by xzilja** [reply]
Makes sense. How is selective resolution handled in pixi? I tried to look for `.resolution` containers / sprites and graphics, but there seems to be no attribute or method for this.

**Reply by bigtimebuddy** [reply]
Good question, there are a couple things if you're looking to optimize your resolution. First, set the Application `resolution` to be the max-resolution for your rendering. Display objects should not exceed this or you're just wasting memory. Here are the places where you can adjust resolution independently:

### Static Rendered Objects
* **BaseTexture** can have an independent resolution. Pixi can handle this automatically if you load images with e.g. `@2x` in the end of the filename. If you're creating Textures or BaseTextures dynamically, you can pass in a resolution at construction time.

### Dynamic Rendered Objects
* **Text** and **BitmapFont** have independent resolutions that you can set, this is because these are dynamically rendered to an intermediary 2D canvas. Setting the resolution basically renderers more pixels.
* **RenderTexture** you can also adjust the resolution using the static method `create` options.
* **Filter** any filter also has a `resolution` that you can reduce to save some memory, helpful with things like Blur.

### Other Resolution Settings
* There are two **default resolution settings** `PIXI.settings.FILTER_RESOLUTION` and `PIXI.settings.RESOLUTION`, in general these can be set before creating your Application/Renderer to set defaults, so you don't need to set the resolution, say, for every `Text` object or filter.
